weatherman.py

```diff
@@ -0,0 +1,178 @@
+
+    def update(self, reading):
+        if self.num_readings == 0:
+            self.highest_temp = self.lowest_temp = self.highest_humidity = reading
+        else:
+            if reading.max_temp > self.highest_temp.max_temp:
+                self.highest_temp = reading
+            if reading.min_temp < self.lowest_temp.min_temp:
+                self.lowest_temp = reading
+            if reading.max_humidity > self.highest_humidity.max_humidity:
+                self.highest_humidity = reading
+
+        self.total_max_temp += reading.max_temp
+        self.total_min_temp += reading.min_temp
+        self.total_mean_humidity += reading.mean_humidity
+        self.num_readings += 1
+
+    def compute_averages(self):
+        if self.num_readings == 0:
+            return None, None, None
+        avg_max_temp = self.total_max_temp / self.num_readings
+        avg_min_temp = self.total_min_temp / self.num_readings
+        avg_mean_humidity = self.total_mean_humidity / self.num_readings
+        return avg_max_temp, avg_min_temp, avg_mean_humidity
+
+
+def read_yearly_data(directory, year):
+    calculations = WeatherCalculations()
+    for month in calendar.month_abbr:
+        if month == "":
+            continue
+        file_path = os.path.join(directory, f"Murree_weather_{year}_{month}.txt")
+        readings = parse_weather_file(file_path)
+        for reading in readings:
+            calculations.update(reading)
+    return calculations
+
+
+def generate_yearly_report(calculations):
+    if calculations.highest_temp and calculations.lowest_temp and calculations.highest_humidity:
+        print(f"Highest: {calculations.highest_temp.max_temp}C on {calculations.highest_temp.date}")
+        print(f"Lowest: {calculations.lowest_temp.min_temp}C on {calculations.lowest_temp.date}")
+        print(f"Humidity: {calculations.highest_humidity.max_humidity}% on {calculations.highest_humidity.date}")
+
+
+def read_monthly_data(directory, year, month):
+    calculations = WeatherCalculations()
+    month_name = datetime(year, month, 1).strftime('%b')
+    file_path = os.path.join(directory, f"Murree_weather_{year}_{month_name}.txt")
+    readings = parse_weather_file(file_path)
+    for reading in readings:
+        calculations.update(reading)
+    return calculations
+
+
+def generate_monthly_report(calculations):
+    avg_max_temp, avg_min_temp, avg_mean_humidity = calculations.compute_averages()
+    print(f"Highest Average Temperature: {avg_max_temp:.0f}C")
+    print(f"Lowest Average Temperature: {avg_min_temp:.0f}C")
+    print(f"Average Mean Humidity: {avg_mean_humidity:.0f}%")
+
+
+def read_monthly_column_data(directory, year, month):
+    month_name = datetime(year, month, 1).strftime("%b")
+    file_path = os.path.join(directory, f"Murree_weather_{year}_{month_name}.txt")
+    readings = parse_weather_file(file_path)
+    return readings
+
+
+def generate_monthly_bar_chart(readings, combined):
+    if readings:
```

Comment:
Having a conditional statement and all things moving a tab ahead will make readability tougher. It's better to handle the negative case to save indentation level and continue with lesser indent. You can also manage to invoke this function by checking readings to save this thing.